### DISCUSSION TITLE
[Feature] Map synthetic (but real!) keys

### DISCUSSION THREAD

**Discussion by kniteli** [houmain/keymapper#143] [Ideas]
Currently, unless I'm missing something, I can't seem to map keys that actually exist, but not on my particular keyboard. For example, I like to map the high F keys for specific vim bindings. I have to do some convoluted stuff with kitty term currently, because keymapper will only map keys that exist on my actual keyboard.

I propose allow output mappings of keys that exist in the [protocol outlined here](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#fixterms-bugs), whether the keyboard currently used has the keys or not. It seems that adoption is becoming widespread and that is probably a safe baseline. Or at least [this table](https://www.toptal.com/developers/keycode/table), regardless of the keys existence on my keyboard.

Basically, this is the difference between mapping synthetic keys that no other programs will understand, such as keymapper's VirtualX keys, and synthetic keys that exist in the globally shared key space.

**Comment by ristomatti** [comment]
The [table](https://www.toptal.com/developers/keycode/table) you mentioned is AFAIK the exact list of named keys supported by Keymapper for either input or output. You don't have to have the key `F18` on your keyboard to be able to output.

I just added this to my config and I then then output `F18` even if the keyboard I'm typing on has only the standard set of F1-F12.

```
F1 >> F18
```

![image](https://github.com/houmain/keymapper/assets/9029939/39e6f2a2-2623-43c6-930e-45c8569c3111)

If this is not what you meant, could you expand on what you'd want to be able to output? Keymapper can use keys outside of that table for both input and output using keycodes. For instance, I can do:

```
F1 >> 448
```

..and then monitor on Keymapper's virtual input device using `evdev`. While monitoring, pressing `F1` on my keyboard will produce this on console:

```
Event: time 1715453625.075648, type 1 (EV_KEY), code 448 (KEY_DEL_EOL), value 1
Event: time 1715453625.075648, -------------- SYN_REPORT ------------
Event: time 1715453625.154676, type 1 (EV_KEY), code 448 (KEY_DEL_EOL), value 0
Event: time 1715453625.154676, -------------- SYN_REPORT ------------
```

The name `KEY_DEV_EOL` is the name given to his keycode in the Linux kernel (there's a link to the code responsible for this on my open pull request #124).

**Reply by ristomatti** [reply]
Below is the complete list of key events Keymapper's virtual device reports to support on Linux. Note that `/dev/keymapper` is a symlink I've configured `udev` to create myself so you won't find it by default. To find the device path, you can just type `sudo evtest` to find what it currently is (it changes each time `keymapper` restarts, even after changing the config).

```
$ sudo evtest /dev/keymapper
Input driver version is 1.0.1
Input device ID: bus 0x3 vendor 0xd1ce product 0x1 version 0x1
Input device name: "Keymapper"
Supported events:
  Event type 0 (EV_SYN)
  Event type 1 (EV_KEY)
    Event code 1 (KEY_ESC)
    Event code 2 (KEY_1)
    Event code 3 (KEY_2)
    Event code 4 (KEY_3)
    Event code 5 (KEY_4)
    Event code 6 (KEY_5)
    Event code 7 (KEY_6)
    Event code 8 (KEY_7)
    Event code 9 (KEY_8)
    Event code 10 (KEY_9)
    Event code 11 (KEY_0)
    Event code 12 (KEY_MINUS)
    Event code 13 (KEY_EQUAL)
    Event code 14 (KEY_BACKSPACE)
    Event code 15 (KEY_TAB)
    Event code 16 (KEY_Q)
    Event code 17 (KEY_W)
    Event code 18 (KEY_E)
    Event code 19 (KEY_R)
    Event code 20 (KEY_T)
    Event code 21 (KEY_Y)
    Event code 22 (KEY_U)
    Event code 23 (KEY_I)
    Event code 24 (KEY_O)
    Event code 25 (KEY_P)
    Event code 26 (KEY_LEFTBRACE)
    Event code 27 (KEY_RIGHTBRACE)
    Event code 28 (KEY_ENTER)
    Event code 29 (KEY_LEFTCTRL)
    Event code 30 (KEY_A)
    Event code 31 (KEY_S)
    Event code 32 (KEY_D)
    Event code 33 (KEY_F)
    Event code 34 (KEY_G)
    Event code 35 (KEY_H)
    Event code 36 (KEY_J)
    Event code 37 (KEY_K)
    Event code 38 (KEY_L)
    Event code 39 (KEY_SEMICOLON)
    Event code 40 (KEY_APOSTROPHE)
    Event code 41 (KEY_GRAVE)
    Event code 42 (KEY_LEFTSHIFT)
    Event code 43 (KEY_BACKSLASH)
    Event code 44 (KEY_Z)
    Event code 45 (KEY_X)
    Event code 46 (KEY_C)
    Event code 47 (KEY_V)
    Event code 48 (KEY_B)
    Event code 49 (KEY_N)
    Event code 50 (KEY_M)
    Event code 51 (KEY_COMMA)
    Event code 52 (KEY_DOT)
    Event code 53 (KEY_SLASH)
    Event code 54 (KEY_RIGHTSHIFT)
    Event code 55 (KEY_KPASTERISK)
    Event code 56 (KEY_LEFTALT)
    Event code 57 (KEY_SPACE)
    Event code 58 (KEY_CAPSLOCK)
    Event code 59 (KEY_F1)
    Event code 60 (KEY_F2)
    Event code 61 (KEY_F3)
    Event code 62 (KEY_F4)
    Event code 63 (KEY_F5)
    Event code 64 (KEY_F6)
    Event code 65 (KEY_F7)
    Event code 66 (KEY_F8)
    Event code 67 (KEY_F9)
    Event code 68 (KEY_F10)
    Event code 69 (KEY_NUMLOCK)
    Event code 70 (KEY_SCROLLLOCK)
    Event code 71 (KEY_KP7)
    Event code 72 (KEY_KP8)
    Event code 73 (KEY_KP9)
    Event code 74 (KEY_KPMINUS)
    Event code 75 (KEY_KP4)
    Event code 76 (KEY_KP5)
    Event code 77 (KEY_KP6)
    Event code 78 (KEY_KPPLUS)
    Event code 79 (KEY_KP1)
    Event code 80 (KEY_KP2)
    Event code 81 (KEY_KP3)
    Event code 82 (KEY_KP0)
    Event code 83 (KEY_KPDOT)
    Event code 84 (?)
    Event code 85 (KEY_ZENKAKUHANKAKU)
    Event code 86 (KEY_102ND)
    Event code 87 (KEY_F11)
    Event code 88 (KEY_F12)
    Event code 89 (KEY_RO)
    Event code 90 (KEY_KATAKANA)
    Event code 91 (KEY_HIRAGANA)
    Event code 92 (KEY_HENKAN)
    Event code 93 (KEY_KATAKANAHIRAGANA)
    Event code 94 (KEY_MUHENKAN)
    Event code 95 (KEY_KPJPCOMMA)
    Event code 96 (KEY_KPENTER)
    Event code 97 (KEY_RIGHTCTRL)
    Event code 98 (KEY_KPSLASH)
    Event code 99 (KEY_SYSRQ)
    Event code 100 (KEY_RIGHTALT)
    Event code 101 (KEY_LINEFEED)
    Event code 102 (KEY_HOME)
    Event code 103 (KEY_UP)
    Event code 104 (KEY_PAGEUP)
    Event code 105 (KEY_LEFT)
    Event code 106 (KEY_RIGHT)
    Event code 107 (KEY_END)
    Event code 108 (KEY_DOWN)
    Event code 109 (KEY_PAGEDOWN)
    Event code 110 (KEY_INSERT)
    Event code 111 (KEY_DELETE)
    Event code 112 (KEY_MACRO)
    Event code 113 (KEY_MUTE)
    Event code 114 (KEY_VOLUMEDOWN)
    Event code 115 (KEY_VOLUMEUP)
    Event code 116 (KEY_POWER)
    Event code 117 (KEY_KPEQUAL)
    Event code 118 (KEY_KPPLUSMINUS)
    Event code 119 (KEY_PAUSE)
    Event code 120 (KEY_SCALE)
    Event code 121 (KEY_KPCOMMA)
    Event code 122 (KEY_HANGUEL)
    Event code 123 (KEY_HANJA)
    Event code 124 (KEY_YEN)
    Event code 125 (KEY_LEFTMETA)
    Event code 126 (KEY_RIGHTMETA)
    Event code 127 (KEY_COMPOSE)
    Event code 128 (KEY_STOP)
    Event code 129 (KEY_AGAIN)
    Event code 130 (KEY_PROPS)
    Event code 131 (KEY_UNDO)
    Event code 132 (KEY_FRONT)
    Event code 133 (KEY_COPY)
    Event code 134 (KEY_OPEN)
    Event code 135 (KEY_PASTE)
    Event code 136 (KEY_FIND)
    Event code 137 (KEY_CUT)
    Event code 138 (KEY_HELP)
    Event code 139 (KEY_MENU)
    Event code 140 (KEY_CALC)
    Event code 141 (KEY_SETUP)
    Event code 142 (KEY_SLEEP)
    Event code 143 (KEY_WAKEUP)
    Event code 144 (KEY_FILE)
    Event code 145 (KEY_SENDFILE)
    Event code 146 (KEY_DELETEFILE)
    Event code 147 (KEY_XFER)
    Event code 148 (KEY_PROG1)
    Event code 149 (KEY_PROG2)
    Event code 150 (KEY_WWW)
    Event code 151 (KEY_MSDOS)
    Event code 152 (KEY_SCREENLOCK)
    Event code 153 (KEY_DIRECTION)
    Event code 154 (KEY_CYCLEWINDOWS)
    Event code 155 (KEY_MAIL)
    Event code 156 (KEY_BOOKMARKS)
    Event code 157 (KEY_COMPUTER)
    Event code 158 (KEY_BACK)
    Event code 159 (KEY_FORWARD)
    Event code 160 (KEY_CLOSECD)
    Event code 161 (KEY_EJECTCD)
    Event code 162 (KEY_EJECTCLOSECD)
    Event code 163 (KEY_NEXTSONG)
    Event code 164 (KEY_PLAYPAUSE)
    Event code 165 (KEY_PREVIOUSSONG)
    Event code 166 (KEY_STOPCD)
    Event code 167 (KEY_RECORD)
    Event code 168 (KEY_REWIND)
    Event code 169 (KEY_PHONE)
    Event code 170 (KEY_ISO)
    Event code 171 (KEY_CONFIG)
    Event code 172 (KEY_HOMEPAGE)
    Event code 173 (KEY_REFRESH)
    Event code 174 (KEY_EXIT)
    Event code 175 (KEY_MOVE)
    Event code 176 (KEY_EDIT)
    Event code 177 (KEY_SCROLLUP)
    Event code 178 (KEY_SCROLLDOWN)
    Event code 179 (KEY_KPLEFTPAREN)
    Event code 180 (KEY_KPRIGHTPAREN)
    Event code 181 (KEY_NEW)
    Event code 182 (KEY_REDO)
    Event code 183 (KEY_F13)
    Event code 184 (KEY_F14)
    Event code 185 (KEY_F15)
    Event code 186 (KEY_F16)
    Event code 187 (KEY_F17)
    Event code 188 (KEY_F18)
    Event code 189 (KEY_F19)
    Event code 190 (KEY_F20)
    Event code 191 (KEY_F21)
    Event code 192 (KEY_F22)
    Event code 193 (KEY_F23)
    Event code 194 (KEY_F24)
    Event code 195 (?)
    Event code 196 (?)
    Event code 197 (?)
    Event code 198 (?)
    Event code 199 (?)
    Event code 200 (KEY_PLAYCD)
    Event code 201 (KEY_PAUSECD)
    Event code 202 (KEY_PROG3)
    Event code 203 (KEY_PROG4)
    Event code 204 (KEY_DASHBOARD)
    Event code 205 (KEY_SUSPEND)
    Event code 206 (KEY_CLOSE)
    Event code 207 (KEY_PLAY)
    Event code 208 (KEY_FASTFORWARD)
    Event code 209 (KEY_BASSBOOST)
    Event code 210 (KEY_PRINT)
    Event code 211 (KEY_HP)
    Event code 212 (KEY_CAMERA)
    Event code 213 (KEY_SOUND)
    Event code 214 (KEY_QUESTION)
    Event code 215 (KEY_EMAIL)
    Event code 216 (KEY_CHAT)
    Event code 217 (KEY_SEARCH)
    Event code 218 (KEY_CONNECT)
    Event code 219 (KEY_FINANCE)
    Event code 220 (KEY_SPORT)
    Event code 221 (KEY_SHOP)
    Event code 222 (KEY_ALTERASE)
    Event code 223 (KEY_CANCEL)
    Event code 224 (KEY_BRIGHTNESSDOWN)
    Event code 225 (KEY_BRIGHTNESSUP)
    Event code 226 (KEY_MEDIA)
    Event code 227 (KEY_SWITCHVIDEOMODE)
    Event code 228 (KEY_KBDILLUMTOGGLE)
    Event code 229 (KEY_KBDILLUMDOWN)
    Event code 230 (KEY_KBDILLUMUP)
    Event code 231 (KEY_SEND)
    Event code 232 (KEY_REPLY)
    Event code 233 (KEY_FORWARDMAIL)
    Event code 234 (KEY_SAVE)
    Event code 235 (KEY_DOCUMENTS)
    Event code 236 (KEY_BATTERY)
    Event code 237 (KEY_BLUETOOTH)
    Event code 238 (KEY_WLAN)
    Event code 239 (KEY_UWB)
    Event code 240 (KEY_UNKNOWN)
    Event code 241 (KEY_VIDEO_NEXT)
    Event code 242 (KEY_VIDEO_PREV)
    Event code 243 (KEY_BRIGHTNESS_CYCLE)
    Event code 244 (KEY_BRIGHTNESS_ZERO)
    Event code 245 (KEY_DISPLAY_OFF)
    Event code 246 (KEY_WIMAX)
    Event code 247 (KEY_RFKILL)
    Event code 248 (KEY_MICMUTE)
    Event code 249 (?)
    Event code 250 (?)
    Event code 251 (?)
    Event code 252 (?)
    Event code 253 (?)
    Event code 254 (?)
    Event code 255 (?)
    Event code 256 (BTN_0)
    Event code 257 (BTN_1)
    Event code 258 (BTN_2)
    Event code 259 (BTN_3)
    Event code 260 (BTN_4)
    Event code 261 (BTN_5)
    Event code 262 (BTN_6)
    Event code 263 (BTN_7)
    Event code 264 (BTN_8)
    Event code 265 (BTN_9)
    Event code 266 (?)
    Event code 267 (?)
    Event code 268 (?)
    Event code 269 (?)
    Event code 270 (?)
    Event code 271 (?)
    Event code 272 (BTN_LEFT)
    Event code 273 (BTN_RIGHT)
    Event code 274 (BTN_MIDDLE)
    Event code 275 (BTN_SIDE)
    Event code 276 (BTN_EXTRA)
    Event code 277 (BTN_FORWARD)
    Event code 278 (BTN_BACK)
    Event code 279 (BTN_TASK)
    Event code 280 (?)
    Event code 281 (?)
    Event code 282 (?)
    Event code 283 (?)
    Event code 284 (?)
    Event code 285 (?)
    Event code 286 (?)
    Event code 287 (?)
    Event code 288 (BTN_TRIGGER)
    Event code 289 (BTN_THUMB)
    Event code 290 (BTN_THUMB2)
    Event code 291 (BTN_TOP)
    Event code 292 (BTN_TOP2)
    Event code 293 (BTN_PINKIE)
    Event code 294 (BTN_BASE)
    Event code 295 (BTN_BASE2)
    Event code 296 (BTN_BASE3)
    Event code 297 (BTN_BASE4)
    Event code 298 (BTN_BASE5)
    Event code 299 (BTN_BASE6)
    Event code 300 (?)
    Event code 301 (?)
    Event code 302 (?)
    Event code 303 (BTN_DEAD)
    Event code 304 (BTN_SOUTH)
    Event code 305 (BTN_EAST)
    Event code 306 (BTN_C)
    Event code 307 (BTN_NORTH)
    Event code 308 (BTN_WEST)
    Event code 309 (BTN_Z)
    Event code 310 (BTN_TL)
    Event code 311 (BTN_TR)
    Event code 312 (BTN_TL2)
    Event code 313 (BTN_TR2)
    Event code 314 (BTN_SELECT)
    Event code 315 (BTN_START)
    Event code 316 (BTN_MODE)
    Event code 317 (BTN_THUMBL)
    Event code 318 (BTN_THUMBR)
    Event code 319 (?)
    Event code 320 (BTN_TOOL_PEN)
    Event code 321 (BTN_TOOL_RUBBER)
    Event code 322 (BTN_TOOL_BRUSH)
    Event code 323 (BTN_TOOL_PENCIL)
    Event code 324 (BTN_TOOL_AIRBRUSH)
    Event code 325 (BTN_TOOL_FINGER)
    Event code 326 (BTN_TOOL_MOUSE)
    Event code 327 (BTN_TOOL_LENS)
    Event code 328 (BTN_TOOL_QUINTTAP)
    Event code 329 (?)
    Event code 330 (BTN_TOUCH)
    Event code 331 (BTN_STYLUS)
    Event code 332 (BTN_STYLUS2)
    Event code 333 (BTN_TOOL_DOUBLETAP)
    Event code 334 (BTN_TOOL_TRIPLETAP)
    Event code 335 (BTN_TOOL_QUADTAP)
    Event code 336 (BTN_GEAR_DOWN)
    Event code 337 (BTN_GEAR_UP)
    Event code 338 (?)
    Event code 339 (?)
    Event code 340 (?)
    Event code 341 (?)
    Event code 342 (?)
    Event code 343 (?)
    Event code 344 (?)
    Event code 345 (?)
    Event code 346 (?)
    Event code 347 (?)
    Event code 348 (?)
    Event code 349 (?)
    Event code 350 (?)
    Event code 351 (?)
    Event code 352 (KEY_OK)
    Event code 353 (KEY_SELECT)
    Event code 354 (KEY_GOTO)
    Event code 355 (KEY_CLEAR)
    Event code 356 (KEY_POWER2)
    Event code 357 (KEY_OPTION)
    Event code 358 (KEY_INFO)
    Event code 359 (KEY_TIME)
    Event code 360 (KEY_VENDOR)
    Event code 361 (KEY_ARCHIVE)
    Event code 362 (KEY_PROGRAM)
    Event code 363 (KEY_CHANNEL)
    Event code 364 (KEY_FAVORITES)
    Event code 365 (KEY_EPG)
    Event code 366 (KEY_PVR)
    Event code 367 (KEY_MHP)
    Event code 368 (KEY_LANGUAGE)
    Event code 369 (KEY_TITLE)
    Event code 370 (KEY_SUBTITLE)
    Event code 371 (KEY_ANGLE)
    Event code 372 (KEY_ZOOM)
    Event code 373 (KEY_MODE)
    Event code 374 (KEY_KEYBOARD)
    Event code 375 (KEY_SCREEN)
    Event code 376 (KEY_PC)
    Event code 377 (KEY_TV)
    Event code 378 (KEY_TV2)
    Event code 379 (KEY_VCR)
    Event code 380 (KEY_VCR2)
    Event code 381 (KEY_SAT)
    Event code 382 (KEY_SAT2)
    Event code 383 (KEY_CD)
    Event code 384 (KEY_TAPE)
    Event code 385 (KEY_RADIO)
    Event code 386 (KEY_TUNER)
    Event code 387 (KEY_PLAYER)
    Event code 388 (KEY_TEXT)
    Event code 389 (KEY_DVD)
    Event code 390 (KEY_AUX)
    Event code 391 (KEY_MP3)
    Event code 392 (KEY_AUDIO)
    Event code 393 (KEY_VIDEO)
    Event code 394 (KEY_DIRECTORY)
    Event code 395 (KEY_LIST)
    Event code 396 (KEY_MEMO)
    Event code 397 (KEY_CALENDAR)
    Event code 398 (KEY_RED)
    Event code 399 (KEY_GREEN)
    Event code 400 (KEY_YELLOW)
    Event code 401 (KEY_BLUE)
    Event code 402 (KEY_CHANNELUP)
    Event code 403 (KEY_CHANNELDOWN)
    Event code 404 (KEY_FIRST)
    Event code 405 (KEY_LAST)
    Event code 406 (KEY_AB)
    Event code 407 (KEY_NEXT)
    Event code 408 (KEY_RESTART)
    Event code 409 (KEY_SLOW)
    Event code 410 (KEY_SHUFFLE)
    Event code 411 (KEY_BREAK)
    Event code 412 (KEY_PREVIOUS)
    Event code 413 (KEY_DIGITS)
    Event code 414 (KEY_TEEN)
    Event code 415 (KEY_TWEN)
    Event code 416 (KEY_VIDEOPHONE)
    Event code 417 (KEY_GAMES)
    Event code 418 (KEY_ZOOMIN)
    Event code 419 (KEY_ZOOMOUT)
    Event code 420 (KEY_ZOOMRESET)
    Event code 421 (KEY_WORDPROCESSOR)
    Event code 422 (KEY_EDITOR)
    Event code 423 (KEY_SPREADSHEET)
    Event code 424 (KEY_GRAPHICSEDITOR)
    Event code 425 (KEY_PRESENTATION)
    Event code 426 (KEY_DATABASE)
    Event code 427 (KEY_NEWS)
    Event code 428 (KEY_VOICEMAIL)
    Event code 429 (KEY_ADDRESSBOOK)
    Event code 430 (KEY_MESSENGER)
    Event code 431 (KEY_DISPLAYTOGGLE)
    Event code 432 (KEY_SPELLCHECK)
    Event code 433 (KEY_LOGOFF)
    Event code 434 (KEY_DOLLAR)
    Event code 435 (KEY_EURO)
    Event code 436 (KEY_FRAMEBACK)
    Event code 437 (KEY_FRAMEFORWARD)
    Event code 438 (KEY_CONTEXT_MENU)
    Event code 439 (KEY_MEDIA_REPEAT)
    Event code 440 (KEY_10CHANNELSUP)
    Event code 441 (KEY_10CHANNELSDOWN)
    Event code 442 (KEY_IMAGES)
    Event code 443 (?)
    Event code 444 (?)
    Event code 445 (?)
    Event code 446 (?)
    Event code 447 (?)
    Event code 448 (KEY_DEL_EOL)
    Event code 449 (KEY_DEL_EOS)
    Event code 450 (KEY_INS_LINE)
    Event code 451 (KEY_DEL_LINE)
    Event code 452 (?)
    Event code 453 (?)
    Event code 454 (?)
    Event code 455 (?)
    Event code 456 (?)
    Event code 457 (?)
    Event code 458 (?)
    Event code 459 (?)
    Event code 460 (?)
    Event code 461 (?)
    Event code 462 (?)
    Event code 463 (?)
    Event code 464 (KEY_FN)
    Event code 465 (KEY_FN_ESC)
    Event code 466 (KEY_FN_F1)
    Event code 467 (KEY_FN_F2)
    Event code 468 (KEY_FN_F3)
    Event code 469 (KEY_FN_F4)
    Event code 470 (KEY_FN_F5)
    Event code 471 (KEY_FN_F6)
    Event code 472 (KEY_FN_F7)
    Event code 473 (KEY_FN_F8)
    Event code 474 (KEY_FN_F9)
    Event code 475 (KEY_FN_F10)
    Event code 476 (KEY_FN_F11)
    Event code 477 (KEY_FN_F12)
    Event code 478 (KEY_FN_1)
    Event code 479 (KEY_FN_2)
    Event code 480 (KEY_FN_D)
    Event code 481 (KEY_FN_E)
    Event code 482 (KEY_FN_F)
    Event code 483 (KEY_FN_S)
    Event code 484 (KEY_FN_B)
    Event code 485 (?)
    Event code 486 (?)
    Event code 487 (?)
    Event code 488 (?)
    Event code 489 (?)
    Event code 490 (?)
    Event code 491 (?)
    Event code 492 (?)
    Event code 493 (?)
    Event code 494 (?)
    Event code 495 (?)
    Event code 496 (?)
    Event code 497 (KEY_BRL_DOT1)
    Event code 498 (KEY_BRL_DOT2)
    Event code 499 (KEY_BRL_DOT3)
    Event code 500 (KEY_BRL_DOT4)
    Event code 501 (KEY_BRL_DOT5)
    Event code 502 (KEY_BRL_DOT6)
    Event code 503 (KEY_BRL_DOT7)
    Event code 504 (KEY_BRL_DOT8)
    Event code 505 (KEY_BRL_DOT9)
    Event code 506 (KEY_BRL_DOT10)
    Event code 507 (?)
    Event code 508 (?)
    Event code 509 (?)
    Event code 510 (?)
    Event code 511 (?)
```
